dask/dataframe/io/parquet/core.py

```diff
@@ -0,0 +1,452 @@
+        metadata, if present). Provide a single field name instead of a list to
+        read in the data as a Series.
+    filters : list
+        List of filters to apply, like ``[('x', '>', 0), ...]``. This implements
+        row-group (partition) -level filtering only, i.e., to prevent the
+        loading of some chunks of the data, and only if relevant statistics
+        have been included in the metadata.
+    index : string, list, False or None (default)
+        Field name(s) to use as the output frame index. By default will be
+        inferred from the pandas parquet file metadata (if present). Use False
+        to read all fields as columns.
+    categories : list, dict or None
+        For any fields listed here, if the parquet encoding is Dictionary,
+        the column will be created with dtype category. Use only if it is
+        guaranteed that the column is encoded as dictionary in all row-groups.
+        If a list, assumes up to 2**16-1 labels; if a dict, specify the number
+        of labels expected; if None, will load categories automatically for
+        data written by dask/fastparquet, not otherwise.
+    storage_options : dict
+        Key/value pairs to be passed on to the file-system backend, if any.
+    engine : {'auto', 'fastparquet', 'pyarrow'}, default 'auto'
+        Parquet reader library to use. If only one library is installed, it
+        will use that one; if both, it will use 'fastparquet'
+    infer_divisions : bool or None (default).
+        By default, divisions are inferred if the read `engine` supports
+        doing so efficiently and the `index` of the underlying dataset is
+        sorted across the individual parquet files. Set to ``True`` to
+        force divisions to be inferred in all cases. Note that this may
+        require reading metadata from each file in the dataset, which may
+        be expensive. Set to ``False`` to never infer divisions.
+
+    Examples
+    --------
+    >>> df = dd.read_parquet('s3://bucket/my-parquet-data')  # doctest: +SKIP
+
+    See Also
+    --------
+    to_parquet
+    """
+    if isinstance(columns, str):
+        df = read_parquet(
+            path,
+            [columns],
+            filters,
+            categories,
+            index,
+            storage_options,
+            engine,
+            gather_statistics,
+        )
+        return df[columns]
+
+    if columns is not None:
+        columns = list(columns)
+
+    name = "read-parquet-" + tokenize(
+        path,
+        columns,
+        filters,
+        categories,
+        index,
+        storage_options,
+        engine,
+        gather_statistics,
+    )
+
+    if isinstance(engine, str):
+        engine = get_engine(engine)
+
+    fs, fs_token, paths = get_fs_token_paths(
+        path, mode="rb", storage_options=storage_options
+    )
+
+    paths = sorted(paths, key=natural_sort_key)  # numeric rather than glob ordering
+
+    meta, statistics, parts = engine.read_metadata(
```

Comment:
I'm not sure I follow.  Is this metadata at the full dataset level or on the per-rowgroup level?  I'm not as familiar with the range index changes.  Does this mean that the parquet libraries have started to store Pandas RangeIndex objects as just metadata rather than actual values?  If so, is this something that we would leave to the parquet libraries to handle on their own?

Reply:
> the parquet libraries have started to store Pandas RangeIndex objects as just metadata rather than actual values

pyarrow does this, yes. It means that the index of each part is automatically set on load (I don't think there's a way to stop it), and that we can know the values of the index in each file before loading, if we "gather_statistics" (but the values would not, as it stands, be in the statistics, since there is no corresponding column stored)

I will need to check if fastparquet handles these correctly (I suspect yes), but it certainly doesn't write them.